### DISCUSSION TITLE
Better persistance for total_daily_energy

### DISCUSSION THREAD

**Discussion by josesa** [emporia-vue-local/esphome#279] [General]
I'm having issues with my setup and looking for suggestion on how to address it.
The Emporia works ok, but the total_daily_energy keeps going to 0 due reboots. I can't tell exactly why the reboots happen, I suspect wifi issues, but I would like to avoid them all together.

Any suggestion on how to make the sum outside of the device? I am pushing data to MQTT and was wondering if I could do the sum in Node-RED or somewhere else.

**Comment by flaviut** [comment]
1. Home Assistant doesn't mind this resetting to 0. The way it works whenever HA sees a decrease in the daily energy value, it assumes it has reset to 0 and starts over on accumulating.
2. [There's a FAQ about wifi connection issues](https://esphome.io/guides/faq.html#wifi-problems), maybe one of those steps will help your issue? There's also the generic advice about making sure your antenna is connected tightly, isn't too close to metal, is outside the panel box, and has a wifi access point reasonably far away.

**Reply by 0n3man** [reply]
@josesa I just did an install and found I had the reboot issue.   While @flaviut indicates it's not a problem because the HA energy screens handles the resets, I wanted these values so I could display numeric values on my modified energy screen, and I couldn't find any way to access the HA values.  I assumed they would be entities, but they are not.  While you can put your mouse over the bars in the energy graphs to get the HA computed values, this isn't very user friendly if you want to look at all the values at once.  

Reliable daily values from the emporia device would give me what I wanted.  I posted about it [here](https://github.com/emporia-vue-local/esphome/discussions/264#discussioncomment-11949000).  I then found your thread, which I was happy to see, as no body responded when I asked if anyone else was seeing reboots in the other thread..  

I just want to give you an update, I don't understand why it fixes the issue for me, but I added the debug code from [here ](https://esphome.io/components/debug.html) to get access to the reboot reason after the box rebooted, With the debug code enable the device doesn't reboot, so I can't verify a connection issue was causing the reboot.  But, I now have reliable daily values coming from the emporia gen3 device.  I'd be interested to know what version device you have.  

@flaviut I'm an old-time programmer and this really looks like a buffer overflow issue where some code is stepping on the heap.  When a change in the address space, that happens with the addition of the extra debug code, makes the issue go away, it's hard to believe this is connection related.  As you're a maintainer, can you suggest some debug steps you would like me to run to get to the root cause of the issue?

**Reply by flaviut** [reply]
@0n3man this integration is only 150 LOC of straightforward C++ code: https://github.com/emporia-vue-local/esphome/blob/dev/esphome/components/emporia_vue/emporia_vue.cpp

The only thing that is potentially sketchy is allocating the i2c message on the stack, since it's about 245 bytes. Other than that, I can't see anything that could be causing an overflow. The rest of the esphome c++ is here if you'd like to review it: https://github.com/esphome/esphome/tree/dev/esphome/components

I really don't have any suggestions on how to approach finding a root cause here unfortunately. Supposedly there's JTAG on the ESP32, but I've never tried using it.

**Reply by 0n3man** [reply]
@flaviut thanks for the feedback.  While the JTAG might be an interesting option, I'd have to acquire another unit, as I'd wouldn't want to have the device apart while on the live wires.  I suspect the issue is probably in the main esphome code base.  I guess I'll look around at that code and see if I can find another way to capture the reboot cause without enabling the debug code.  If I capture the reboot cause it will give me a way to focus my effort.  Out of curiosity do you also see the reboot issue and if so have you tried to use the debug component to capture the cause?

**Comment by t0ny-peng** [comment]
You should use the `utility_meter` integration. It as an option to support sensors that resets to 0 on boot, e.g., this VUE2 sensor.
